src/Common/UnreleasableLRUCache.h

```diff
@@ -0,0 +1,283 @@
+        return cells.size();
+    }
+
+    size_t maxSize() const
+    {
+        return max_size;
+    }
+
+    void reset()
+    {
+        std::lock_guard lock(mutex);
+        queue.clear();
+        cells.clear();
+        current_size = 0;
+        hits = 0;
+        misses = 0;
+    }
+
+    virtual ~UnreleasableLRUCache() {}
+
+protected:
+    using LRUQueue = std::list<Key>;
+    using LRUQueueIterator = typename LRUQueue::iterator;
+
+    struct Cell
+    {
+        MappedPtr value;
+        size_t size;
+        LRUQueueIterator queue_iterator;
+    };
+
+    using Cells = std::unordered_map<Key, Cell, HashFunction>;
+
+    Cells cells;
+
+    mutable std::mutex mutex;
+private:
+
+    LRUQueue queue;
+
+    /// Total weight of values.
+    size_t current_size = 0;
+    const size_t max_size;
+    const size_t max_elements_size;
+
+    std::atomic<size_t> hits {0};
+    std::atomic<size_t> misses {0};
+
+    WeightFunction weight_function;
+    EvictPolicy evict_policy;
+
+    MappedPtr getImpl(const Key & key, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
+    {
+        auto it = cells.find(key);
+        if (it == cells.end())
+        {
+            return MappedPtr();
+        }
+
+        Cell & cell = it->second;
+
+        /// Move the key to the end of the queue. The iterator remains valid.
+        queue.splice(queue.end(), queue, cell.queue_iterator);
+
+        return cell.value;
+    }
+
+    bool setImpl(const Key & key, const MappedPtr & mapped, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
+    {
+        auto [it, inserted] = cells.emplace(std::piecewise_construct,
+            std::forward_as_tuple(key),
+            std::forward_as_tuple());
+
+        Cell & cell = it->second;
+
+        if (inserted)
+        {
+            auto weight = mapped ? weight_function(*mapped) : 0;
+            if (!removeOverflow(weight))
+            {
+                // cannot insert this new value
+                cells.erase(it);
+                return false;
+            }
+
+            try
+            {
+                cell.queue_iterator = queue.insert(queue.end(), key);
+            }
+            catch (...)
+            {
+                cells.erase(it);
+                throw;
+            }
+        }
+        else
+        {
+            if (evict_policy.canRelease(*cell.value) != CacheEvictStatus::CAN_EVITCT)
+            {
+                // the old value is not releasable
+                return false;
+            }
+            evict_policy.release(*cell.value);
+            current_size -= cell.size;
+            queue.splice(queue.end(), queue, cell.queue_iterator);
+        }
+
+        cell.value = mapped;
+        cell.size = cell.value ? weight_function(*cell.value) : 0;
+        current_size += cell.size;
+
+        removeOverflow(0);
+        return true;
+    }
+
+    // Or make your own implementation
+    virtual bool removeOverflow(size_t more_size)
```

Comment:
Why not just inherit from `LRUCache`? It is definitely not ok to copy paste all implementation from `LRUCache` and add some extensions.
I see your answer might be - you added this class to add a `typename EvictPolicy = TrivialUnreleasableLRUCacheEvitPolicy<TMapped>`, but seems like this logic can be put in the inherited class and no need in one more template.

Reply:
Ok, I see some other places. But anyway this is not ok, need to try to extend in some different way. Without copying full implementation. May be try to make changes to initial class, and will see if the extension fits there.
But I would prefer to extend only though inheriting.

Reply:
changed